node/core/chain-selection/src/db_backend/v1.rs

```diff
@@ -0,0 +1,668 @@
+			block_number: x.block_number,
+			block_hash: x.block_hash,
+		}
+	}
+}
+
+impl From<LeafEntry> for crate::LeafEntry {
+	fn from(x: LeafEntry) -> crate::LeafEntry {
+		crate::LeafEntry {
+			weight: x.weight,
+			block_number: x.block_number,
+			block_hash: x.block_hash,
+		}
+	}
+}
+
+#[derive(Encode, Decode)]
+struct LeafEntrySet {
+	inner: Vec<LeafEntry>,
+}
+
+impl From<crate::LeafEntrySet> for LeafEntrySet {
+	fn from(x: crate::LeafEntrySet) -> Self {
+		LeafEntrySet {
+			inner: x.inner.into_iter().map(Into::into).collect(),
+		}
+	}
+}
+
+impl From<LeafEntrySet> for crate::LeafEntrySet {
+	fn from(x: LeafEntrySet) -> crate::LeafEntrySet {
+		crate::LeafEntrySet {
+			inner: x.inner.into_iter().map(Into::into).collect(),
+		}
+	}
+}
+
+#[derive(Debug, Encode, Decode, Clone, PartialEq)]
+struct BlockEntry {
+	block_hash: Hash,
+	block_number: BlockNumber,
+	parent_hash: Hash,
+	children: Vec<Hash>,
+	viability: ViabilityCriteria,
+	weight: BlockWeight,
+}
+
+impl From<crate::BlockEntry> for BlockEntry {
+	fn from(x: crate::BlockEntry) -> Self {
+		BlockEntry {
+			block_hash: x.block_hash,
+			block_number: x.block_number,
+			parent_hash: x.parent_hash,
+			children: x.children,
+			viability: x.viability.into(),
+			weight: x.weight,
+		}
+	}
+}
+
+impl From<BlockEntry> for crate::BlockEntry {
+	fn from(x: BlockEntry) -> crate::BlockEntry {
+		crate::BlockEntry {
+			block_hash: x.block_hash,
+			block_number: x.block_number,
+			parent_hash: x.parent_hash,
+			children: x.children,
+			viability: x.viability.into(),
+			weight: x.weight,
+		}
+	}
+}
+
+/// Configuration for the database backend.
+#[derive(Debug, Clone, Copy)]
+pub struct Config {
+	/// The column where block metadata is stored.
+	pub col_data: u32,
+}
+
+/// The database backend.
+pub struct DbBackend {
+	inner: Arc<dyn KeyValueDB>,
+	config: Config,
+}
+
+impl DbBackend {
+	/// Create a new [`DbBackend`] with the supplied key-value store and
+	/// config.
+	pub fn new(db: Arc<dyn KeyValueDB>, config: Config) -> Self {
+		DbBackend {
+			inner: db,
+			config,
+		}
+	}
+}
+
+impl Backend for DbBackend {
+	fn load_block_entry(&self, hash: &Hash) -> Result<Option<crate::BlockEntry>, Error> {
+		load_decode::<BlockEntry>(
+			&*self.inner,
+			self.config.col_data,
+			&block_entry_key(hash),
+		).map(|o| o.map(Into::into))
+	}
+
+	fn load_leaves(&self) -> Result<crate::LeafEntrySet, Error> {
+		load_decode::<LeafEntrySet>(
+			&*self.inner,
+			self.config.col_data,
+			LEAVES_KEY,
+		).map(|o| o.map(Into::into).unwrap_or_default())
+	}
+
+	fn load_stagnant_at(&self, timestamp: crate::Timestamp) -> Result<Vec<Hash>, Error> {
+		load_decode::<Vec<Hash>>(
+			&*self.inner,
+			self.config.col_data,
+			&stagnant_at_key(timestamp.into()),
+		).map(|o| o.unwrap_or_default())
+	}
+
+	fn load_stagnant_at_up_to(&self, up_to: crate::Timestamp)
+		-> Result<Vec<(crate::Timestamp, Vec<Hash>)>, Error>
+	{
+		let stagnant_at_iter = self.inner.iter_with_prefix(
```

Comment:
just a note, that this adds another reliance on `iter_with_prefix` which needs to be removed if we're to migrate off `kvdb` interface

Reply:
Right. I think that's fine for now based on discussions in Element